packages/swingset-liveslots/src/virtualReferences.js

```diff
@@ -500,26 +500,52 @@ export function makeVirtualReferenceManager(
+   * or the ephemeral voAwareWeakMap/Set that we impose upon userspace
+   * as "WeakMap/WeakSet". The collection can be used to query whether
+   * a future specimen matches the original or not, without holding
+   * onto the original.
    *
-   * This 'vrefRecognizers' is a Map from those vrefs to the set of
-   * recognizing weak collections, for virtual keys and non-virtual
-   * collections. Specifically, the vrefs correspond to imported
-   * Presences or virtual-object Representatives (Remotables do not
-   * participate: they are keyed by the actual Remotable object, not
-   * its vref). The collections are either a VirtualObjectAwareWeakMap
-   * or a VirtualObjectAwareWeakSet. We remove the entry when the key
-   * is removed from the collection, and when the entire collection is
-   * deleted.
+   * We need "recognition records" to map from the vref to the
+   * collection that can recognize it. When the vref is retired, we
+   * use the record to find all the collections from which we need to
+   * delete entries, so we can release the matching values. This might
+   * happen because the vref was for a Presence and the kernel just
+   * told us the upstream vat has deleted it (dispatch.retireImports),
+   * or because it was for a locally-managed object (an ephemeral
+   * Remotable or a virtual/durable Representative) and we decided to
+   * delete it.
    *
-   * It is critical that each collection have exactly one recognizer that is
-   * unique to that collection, because the recognizers themselves will be
-   * tracked by their object identities, but the recognizer cannot be the
-   * collection itself else it would prevent the collection from being garbage
-   * collected.
+   * The virtual/durable collections track their "recognition records"
+   * in the vatstore, in keys like "vom.ir.${vref}|${collectionID}".
+   * These records do not contribute to our RAM usage.
+   *
+   * voAwareWeakMap and voAwareWeakSet store their recognition records
+   * in RAM, using this Map named 'vrefRecognizers'. Each key is a
+   * vref, and the value is a Set of recognizers. Each recognizer is
+   * the internal 'virtualObjectMap' in which the collection maps from
+   * vref to value. These in-RAM collections only use virtualObjectMap
+   * to track Presence-style (imports) and Representative-style
+   * (virtual/durable) vrefs: any Remotable-style keys are stored in
+   * the collection's internal (real) WeakMap under the Remotable
+   * object itself (because the engine handles the bookkeeping, and
+   * there is no virtual data in the value that we need to clean up at
+   * deletion time).
+   *
+   * It is critical that each collection have exactly one recognizer
+   * that is unique to that collection, because the recognizers
+   * themselves will be tracked by their object identities, but the
+   * recognizer cannot be the collection itself else it would prevent
+   * the collection from being garbage collected.
```

Comment:
I don't quite understand this paragraph

Reply:
Yeah, that's leftover from the original implementation, the idea is that each voAwareWeakSet needs a distinct `recognizer`, they can't e.g. share one among the whole class. And, the recognizer is not the same thing as the `voAwareWeakSet` itself, because the strong reference we keep to the recognizer would prevent the `voAwareWeakSet` from being freed. So it must be 1-to-1 but not identical.

Reply:
replaced with:
```
   * Each voAwareWeakMap/Set must have a distinct recognizer, so we
   * can remove the key from the right ones. The recognizer is held
   * strongly by the recognition record, so it must not be the
   * voAwareWeakMap/Set itself (which would inhibit GC).
```